framework/rest/DataFilter.php

```diff
@@ -0,0 +1,809 @@
+    private $_filter;
+    /**
+     * @var Model|array|string|callable model to be used for filter attributes validation.
+     */
+    private $_searchModel;
+    /**
+     * @var array list of search attribute types in format: attributeName => type
+     */
+    private $_searchAttributeTypes;
+
+
+    /**
+     * @return mixed raw filter value.
+     */
+    public function getFilter()
+    {
+        return $this->_filter;
+    }
+
+    /**
+     * @param mixed $filter raw filter value.
+     */
+    public function setFilter($filter)
+    {
+        $this->_filter = $filter;
+    }
+
+    /**
+     * @return Model model instance.
+     * @throws InvalidConfigException on invalid configuration.
+     */
+    public function getSearchModel()
+    {
+        if (!is_object($this->_searchModel) || $this->_searchModel instanceof \Closure) {
+            $model = Yii::createObject($this->_searchModel);
+            if (!$model instanceof Model) {
+                throw new InvalidConfigException('`' . get_class($this) . '::$model` should be an instance of `' . Model::className() . '` or its DI compatible configuration.');
+            }
+            $this->_searchModel = $model;
+        }
+        return $this->_searchModel;
+    }
+
+    /**
+     * @param Model|array|string|callable $model model instance or its DI compatible configuration.
+     * @throws InvalidConfigException on invalid configuration.
+     */
+    public function setSearchModel($model)
+    {
+        if (is_object($model)) {
+            if (!$model instanceof Model && !$model instanceof \Closure) {
+                throw new InvalidConfigException('`' . get_class($this) . '::$model` should be an instance of `' . Model::className() . '` or its DI compatible configuration.');
+            }
+        }
+        $this->_searchModel = $model;
+    }
+
+    /**
+     * @return array search attribute type map.
+     */
+    public function getSearchAttributeTypes()
+    {
+        if ($this->_searchAttributeTypes === null) {
+            $this->_searchAttributeTypes = $this->detectSearchAttributeTypes();
+        }
+        return $this->_searchAttributeTypes;
+    }
+
+    /**
+     * @param array|null $searchAttributeTypes search attribute type map.
+     */
+    public function setSearchAttributeTypes($searchAttributeTypes)
+    {
+        $this->_searchAttributeTypes = $searchAttributeTypes;
+    }
+
+    /**
+     * Composes default value for [[searchAttributeTypes]] from the [[searchModel]] validation rules.
+     * @return array attribute type map.
+     */
+    protected function detectSearchAttributeTypes()
+    {
+        $model = $this->getSearchModel();
+
+        $attributeTypes = [];
+        foreach ($model->activeAttributes() as $attribute) {
+            $attributeTypes[$attribute] = self::TYPE_STRING;
+        }
+
+        foreach ($model->getValidators() as $validator) {
+            $type = null;
+            if ($validator instanceof BooleanValidator) {
```

Comment:
I'd move this part to a separate protected method to make it easier to adjust for custom validators.

Reply:
Disagreed.
It is meant to override a whole `detectSearchAttributeTypes()` ot provide some sophosticated type detecttion mechanism. Particular implementation may not process validators at all.

Reply:
Adding custom validator does not lead to any sophisticated type detection logic. It is still a map: class to type. I don't think that overriding the whole method (including logic) is a right way for filter configuration.

Moreover, current approach increases formal cyclomatic complexity.

Reply:
Repeat once more: method `detectSearchAttributeTypes()` as a whole determines the search attribute types. The actual algorithm of its detection may be **different**.
For example: you may want to get types from DB table schema, or may want to parse PHPDoc searching for type-hints.
It is only a single particular implementation, which relies on model validators.

If I add extra method for validators processing, it will become useless and never will be invoked in case `detectSearchAttributeTypes()` rewritten to use DB table schema.

> I don't think that overriding the whole method (including logic) is a right way for filter configuration.

How is method override related to filter configuration? You can set `$searchAttributeTypes` via `setSearchAttributeTypes()` to whatever you like. Method `detectSearchAttributeTypes()` invoked only in case there are `$searchAttributeTypes` explicitly configured.